### PR TITLE
feat: add recoverer algorithms

### PR DESCRIPTION
**Pull request by minettekaum**
## Description
Added PERP (Parameter-Efficient Recovery Protocol) recovery algorithms to restore model quality after compression. The implementation includes multiple variants for both text-to-image and text-to-text models, supporting LoRA adapters and in-place finetuning mechanisms.

### Core PERP Algorithms
- **Main algorithm classes**: `src/pruna/algorithms/perp.py`
  - `TextToImagePERP` - General-purpose recovery for text-to-image models
  - `TextToImageInPlacePERP` - In-place variant without LoRA (faster inference)
  - `TextToImageLoRA` - LoRA-only variant for text-to-image
  - `TextToTextPERP` - Recovery for text-to-text models (LLMs)
  - `TextToTextInPlacePERP` - In-place variant for text-to-text
  - `TextToTextLoRA` - LoRA-only variant for text-to-text

- **Distillation variants**: `src/pruna/algorithms/distillation_perp.py`
  - `TextToImagePERPDistillation` - PERP with distillation for text-to-image
  - `TextToImageInPlacePERPDistillation` - In-place distillation variant
  - `TextToImageLoraDistillation` - LoRA distillation variant

### Core Recovery Infrastructure
- **Base recoverer**: `src/pruna/algorithms/global_utils/recovery/perp_recoverer.py`
  - `PERPRecoverer` - Base class implementing core recovery logic with adapter management, finetuning orchestration, and scheduler handling

- **Adapter implementations**: `src/pruna/algorithms/global_utils/recovery/adapters/`
  - `norm.py` - NormAdapter for in-place norm finetuning
  - `bias.py` - BiasAdapter for in-place bias finetuning
  - `head.py` - HeadAdapter for head finetuning (text-to-text models)
  - `lora.py` - LoraAdapter using HuggingFace PEFT
  - `utils.py` - Adapter utilities and parameter freezing

- **Finetuner implementations**: `src/pruna/algorithms/global_utils/recovery/finetuners/`
  - `text_to_image_finetuner.py` - TextToImageFinetuner with training loop, loss computation, and scheduler integration
  - `text_to_text_finetuner.py` - TextToTextFinetuner for LLM recovery
  - `text_to_image_distiller.py` - TextToImageDistiller for distillation-based recovery
  - `diffusers/scheduler_interface.py` - Training scheduler utilities (get_training_scheduler, sample_timesteps, add_noise, get_target)
  - `diffusers/pack_and_predict.py` - Pipeline-specific prediction functions for SD, SDXL, Sana, Flux
  - `diffusers/utils.py` - Diffusers pipeline utilities (denoiser extraction, prompt encoding, device management)
  - `diffusers/distillation_arg_utils.py` - Utilities for distillation argument handling

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
- **Text-to-image PERP tests**: `tests/algorithms/testers/tti_perp.py`
  - Tests for `TextToImagePERP` algorithm with various model types

- **Text-to-image in-place PERP tests**: `tests/algorithms/testers/tti_inplace_perp.py`
  - Tests for `TextToImageInPlacePERP` (no LoRA variant)

- **Text-to-image LoRA tests**: `tests/algorithms/testers/tti_lora.py`
  - Tests for `TextToImageLoRA` (LoRA-only variant)

- **Text-to-text PERP tests**: `tests/algorithms/testers/ttt_perp.py`
  - Tests for `TextToTextPERP` algorithm with LLM models

- **Text-to-text in-place PERP tests**: `tests/algorithms/testers/ttt_inplace_perp.py`
  - Tests for `TextToTextInPlacePERP` (no LoRA variant)

- **Text-to-text LoRA tests**: `tests/algorithms/testers/ttt_lora.py`
  - Tests for `TextToTextLoRA` (LoRA-only variant)

- **Distillation PERP tests**: `tests/algorithms/testers/tti_distillation_perp.py`
  - Tests for `TextToImagePERPDistillation` algorithm

- **In-place distillation PERP tests**: `tests/algorithms/testers/tti_distillation_inplace_perp.py`
  - Tests for `TextToImageInPlacePERPDistillation` algorithm

- **LoRA distillation PERP tests**: `tests/algorithms/testers/tti_distillation_lora.py`
  - Tests for `TextToImageLoraDistillation` algorithm

- **Test utilities**: `tests/algorithms/testers/utils.py`
  - `restrict_recovery_time()` - Utility to limit training time for testing
  - `replace_datamodule_with_distillation_datamodule()` - Helper for distillation tests
  - `get_model_sparsity()` - Utility to measure model sparsity

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by gsprochette** [comment]
@cursor review

File: src/pruna/algorithms/global_utils/recovery/adapters/utils.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+    Returns
+    -------
+    int
+        The number of parameters that were activated.
+    int
+        The number of parameters found that match the given name but were not trainable.
+    """
+    if len(target_modules) == 0:
+        return
```
Comment (gsprochette) [inline src/pruna/algorithms/global_utils/recovery/adapters/utils.py]:
```suggestion
        return (0, 0)
```

File: src/pruna/algorithms/global_utils/recovery/adapters/head.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        # find head from type and name
+        model_heads = [
+            component
+            for comp_name, component in inspect.getmembers(model)
+            if isinstance(component, torch.nn.Linear) and "head" in comp_name.lower()
+        ]
+        if len(model_heads) != 1:
+            # = 0: model with no head, e.g. diffusers
+            # > 1: model with multiple heads, e.g. for localization, not currently supported
+            model_head_names = [h[0] for h in model_heads]  # type: ignore[index]
```
[the remark this answers is not in the document]
Comment (gsprochette) [reply src/pruna/algorithms/global_utils/recovery/adapters/head.py]:
```suggestion
            model_head_names = [
                comp_name
                for comp_name, component in inspect.getmembers(model)
                if isinstance(component, torch.nn.Linear) and "head" in comp_name.lower()
            ]
```

File: src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            The text field to use.
+
+        Returns
+        -------
+        tuple[Dataset, Optional[str]]
+            The formatted dataset and dataset text field.
+            A text field is only provided if the dataset is a huggingface dataset not yet tokenized.
+        """
+        column_names = dataset.column_names  # type: ignore[union-attr]
+        if hasattr(dataset, "column_names") and "input_ids" in column_names:
```
[the remark this answers is not in the document]
Comment (gsprochette) [reply src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py]: This is a reasonable comment, here is my takedown:
- this problem is 100% due to trying to handle both torch datasets and HF datasets. The handling of torch datasets is a bit slow so one option is to drop that support but it feels very restrictive
- the other option is to reorder, and properly check for type with:
  - isinstance(dataset, Dataset) on one side, only compute the column names in this case and that should help with typing as well.
  - isinstance(dataset, torch.utils.data.Dataset) on the other side this handles only an (input, output) case for next token prediction and converts it to (extendended_inputs,), but we should also handle a dataset already in this format I think

File: src/pruna/algorithms/global_utils/recovery/finetuners/text_to_image_distiller.py [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+            smash_config["warmup_steps"],
+            smash_config["weight_decay"],
+            lora_r,
+            recoverer,
+            smash_config["use_cpu_offloading"],
+            pipeline_kwargs=smash_config.data.pipeline_kwargs,
+        )
+        # make directory for logs and checkpoints
+        model_path = pathlib.Path(smash_config.cache_dir) / "recovery"
+        model_path.mkdir(parents=True)
```
[the remark this answers is not in the document]
Comment (gsprochette) [reply src/pruna/algorithms/global_utils/recovery/finetuners/text_to_image_distiller.py]:
```suggestion
        model_path.mkdir(exist_ok=True, parents=True)
```

File: src/pruna/data/diffuser_distillation_data_module.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+            The dataloader to use to prepare the dataset.
+        subdir_name : str
+            The name of the subdirectory to save the dataset to, in ["train", "val", "test"].
+
+        Returns
+        -------
+        List[str]
+            The filenames of the dataset.
+        """
+        Path(self.save_path / subdir_name).mkdir(exist_ok=True, parents=True)
```
Comment (gsprochette) [inline src/pruna/data/diffuser_distillation_data_module.py:175]: Adding this None case and extending the type hint should take care of the bugbot comment below. Maybe failing more strongly would be a better choice though.
```suggestion
        if dataloader is None:
            pruna_logger.warning(f"Missing dataloader for {subdir_name} data")
            return None
        Path(self.save_path / subdir_name).mkdir(exist_ok=True, parents=True)
```

File: src/pruna/data/diffuser_distillation_data_module.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        self.collection_helper.disable()
+
+        # pipeline should not be needed by this module anymore, so we drop the reference so it can be deleted elsewhere
+        self.collection_helper = None
+        self.pipeline = None
+        return train_filenames, val_filenames, test_filenames
+
+    def _prepare_one_dataset(
+        self,
+        dataloader: Optional[DataLoader],
```
Comment (gsprochette) [inline src/pruna/data/diffuser_distillation_data_module.py:154]:
```suggestion
        dataloader: Optional[DataLoader] | None,
```

File: src/pruna/data/diffuser_distillation_data_module.py [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+import torch
+from torch.utils.data import DataLoader, Dataset
+from tqdm import tqdm
+
+from pruna.algorithms.global_utils.recovery.finetuners.diffusers.distillation_arg_utils import (
+    get_latent_extractor_fn,
+)
+from pruna.algorithms.global_utils.recovery.finetuners.diffusers.utils import get_denoiser_attr
+from pruna.data.pruna_datamodule import PrunaDataModule
+
```
Comment (gsprochette) [inline src/pruna/data/diffuser_distillation_data_module.py:32]:
```suggestion
from pruna.logging.logger import pruna_logger
```

**Review by gsprochette — CHANGES_REQUESTED** [review]
I left some code suggestions fixing the bugbot slander.
There's one specific improvement that can be made on next-token-prediction dataset handling which is very crude. Let me know if you need me to update the function.
If you accept the small changes it's an instant approve on my side :)

File: src/pruna/algorithms/global_utils/recovery/adapters/head.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (simlang) [reply src/pruna/algorithms/global_utils/recovery/adapters/head.py]: a bit hacky, but okay for me, or we also collect the name in model_heads in line 71, then we don't have to go through the model twice

**Review by simlang — APPROVED** [review]
Nothing to add, thank you! 
LGTM 🚀

File: src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py [hunk 8]
```diff
@@ -220,37 +220,93 @@ def _format_dataset_for_causal_lm(
...
+                # remove all other columns, otherwise HF's Trainer tends to infer the wrong task from those columns
+                removed_columns = [col for col in column_names if col not in [text_field, "input_ids"]]
+                return dataset.remove_columns(removed_columns), None
+            elif text_field in column_names:
+                # raw dataset with text field
+                # remove all other columns, otherwise HF's Trainer tends to infer the wrong task from those columns
+                removed_columns = [col for col in column_names if col != text_field]
+                return dataset.remove_columns(removed_columns), text_field
+            else:
+                pruna_logger.error(
```
Comment (gsprochette) [inline src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py]: This same error message appears multiple time, can we factorize it?

File: src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py [hunk 9]
```diff
@@ -220,37 +220,93 @@ def _format_dataset_for_causal_lm(
...
+                    " - (input, label) format for next token prediction,\n"
+                    " - (extended_inputs,) format for next token prediction."
+                )
+                raise ValueError(f"Expected a HuggingFace dataset with text or input_ids fields for LoRA recovery but got: {dataset}")
 
-            dataset = Dataset.from_generator(data_generator)
-            return dataset, None
-        else:
+        # Handle PyTorch Dataset
+        if isinstance(dataset, torch.utils.data.Dataset):
```
Comment (gsprochette) [inline src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py]: elif ?

File: src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py [hunk 10]
```diff
@@ -220,37 +220,93 @@ def _format_dataset_for_causal_lm(
...
+                            sample = dataset[idx]
+                            input_ids = sample[0] if isinstance(sample, (tuple, list)) else sample
+                            if not isinstance(input_ids, torch.Tensor):
+                                input_ids = torch.tensor(input_ids)
+                            attention_mask = torch.ones_like(input_ids)
+                            yield {"input_ids": input_ids, "attention_mask": attention_mask}
+                    return Dataset.from_generator(data_generator), None
+            
+            # Check if in (input, label) format for next token prediction
+            if isinstance(first_sample, (tuple, list)) and len(first_sample) == 2:
```
Comment (gsprochette) [inline src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py]: elif ?

File: src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py [hunk 11]
```diff
@@ -220,37 +220,93 @@ def _format_dataset_for_causal_lm(
...
+
+        # Unknown dataset type
+        pruna_logger.error(
+            "The dataset provided for recovery is not compatible. Accepted format include:\n"
+            " - huggingface datasets with a text field,\n"
+            " - huggingface datasets with an input_ids field,\n"
+            " - (input, label) format for next token prediction,\n"
+            " - (extended_inputs,) format for next token prediction."
+        )
+        raise ValueError(f"Expected a Dataset or torch.utils.data.Dataset for LoRA recovery but got: {type(dataset)}")
```
Comment (gsprochette) [inline src/pruna/algorithms/global_utils/recovery/finetuners/text_to_text_finetuner.py]: Dataset is not super explicit here, `datasets.Dataset` be better to differenciate if from torch.utils.data.Dataset
